### DISCUSSION TITLE
get the installed path of package rather than module

### DISCUSSION THREAD

**Discussion by zhouzaida** [pypa/setuptools#2745] [Q&A] [answered]
I implement a helper function to get the installed path of a package rather than a module.
However, the performance is so slow.
```python
import os.path as osp
from pkg_resources import get_distribution

def get_installed_path(package: str) -> str:
    pkg = get_distribution(package)
    possible_path = osp.join(pkg.location, package)
    if osp.exists(possible_path):
        return possible_path
    else:
        if pkg.has_metadata('top_level.txt'):
            module_name = pkg.get_metadata('top_level.txt').split('\n')[0]
            return osp.join(pkg.location, module_name)
        else:
            raise ValueError(f'can not get the installed path of {package}')

>>> get_installed_path('mysql-python')
'.../lib/python3.7/site-packages/MySQLdb'
>>> get_installed_path('python-memcached')
'.../lib/python3.7/site-packages/memcache'
```

**Comment by webknjaz** [comment, marked as answer]
Is there a question somewhere in this statement?

The only thing I'll mention is that today it's best to use `importlib.metadata` in most places where you'd use `pkg_resources` in the past.

**Reply by zhouzaida** [reply]
Thanks for your quick reply. Yet, the above code has an issue and I fixed it.

In fact, the name of a package is not always the same as its module name(https://stackoverflow.com/questions/11453866/given-the-name-of-a-python-package-what-is-the-name-of-the-module-to-import).

If we want to get the installed path of a package, we should know the module name to concatenate the pkg.location and module_name.

In addition, importlib.metadata is a new feature of python3.8 but my python is 3.6

**Reply by zhouzaida** [reply]
I provide two cases here.

> I implement a helper function to get the installed path of a package rather than a module.
> However, the performance is so slow.
> 
> ```python
> import os.path as osp
> from pkg_resources import get_distribution
> 
> def get_installed_path(package: str) -> str:
>     pkg = get_distribution(package)
>     possible_path = osp.join(pkg.location, package)
>     if osp.exists(possible_path):
>         return possible_path
>     else:
>         if pkg.has_metadata('top_level.txt'):
>             module_name = pkg.get_metadata('top_level.txt').split('\n')[0]
>             return osp.join(pkg.location, module_name)
>         else:
>             raise ValueError(f'can not get the installed path of {package}')
> 
> >>> get_installed_path('mysql-python')
> '.../lib/python3.7/site-packages/MySQLdb'
> >>> get_installed_path('python-memcached')
> '.../lib/python3.7/site-packages/memcache'
> ```

**Reply by webknjaz** [reply]
> If we want to get the installed path of a package, we should know the module name to concatenate the pkg.location and module_name.

True.

> In addition, importlib.metadata is a new feature of python3.8 but my python is 3.6

Yes. That's why a compat project called `importlib_metadata` exists on PyPI. Use it as a fallback.

**Reply by zhouzaida** [reply]
Thanks very much. I have a try

**Reply by zhouzaida** [reply]
Thanks very much. It is very helpful.
